### PR TITLE
Consolidate ORTModule logging

### PR DESCRIPTION
**Pull request by pengwa**
### Consolidate ORTModule logging

There are few improvements for ORTModule loggings:
- All ORTModule logging are used logger that is initialized in `ortmodule.py`.
- Manage all export logs same way, e.g. use ` _logger.suppress_os_stream_output(log_level=self._debug_options.logging.log_level)` to control exporting related logs suppressing or not. If any warning or errors suppressed, `self._warning_log_detected_during_export` will be set to True, then when we log ORTModule feature matrix, we will also told users there are logs suppressed.
- Downgrade some warnings. We had some warnings for years, and looks many models have them by default, no action we actually can take, so downgrade them to make user logging cleaner.
- PyTorch export requires update of custom export function signature changes, otherwise, _symbolic_context_handler complains with warnings, so update custom export function adaption for version >=1.13 PyTorch.
- Add ORTModule feature matrix summary, **this is supposed to be only places users see our logs by default** (unless they use INFO or VERBOSE). Features ON/OFF states are shown clearly to them in case they want to try some features in OFF states. This logs only shows up in rank 0 (if there are multiple rank), the intention is we want user to see a useful and clean output from ORTModule by default. The outputs shown as below:

image

image

- `reinitialize_ortmodule` in util.py is only used by ortmodule.py, moving it into ortmodule.py, then utils takes no dependency on `orttraining/orttraining/python/training/ortmodule/_custom_op_symbolic_registry.py`, then `_custom_op_symbolic_registry.py` can call functions defined in utils.py (without recursively include).

### CODE REVIEW

**Review by baijumeswani — DISMISSED** [review]

File: orttraining/orttraining/python/training/ortmodule/_custom_op_symbolic_registry.py [hunk 1]
```diff
@@ -38,20 +41,43 @@ def pytorch_type_to_onnx(scalar_type: str) -> torch.onnx.TensorProtoDataType:
...
 
-def wrap_custom_export_function(original_func):
-    # Starting from PyTorch 1.11, there has been a change to symbolic function signature
-    # in terms of how additional context is accessed. More info at
-    # https://github.com/pytorch/pytorch/blob/6b02648479d3615fa3260961e24f38dd0f22da94/torch/onnx/symbolic_helper.py#L48
-    # This code can be cleaned up once support for PyTorch version < 1.11 is dropped.
-    try:
+def wrap_custom_export_function(original_func: Callable) -> Callable:
+    """
+    This function is to wrap the custom export function to make sure it can be used by different versions of PyTorch.
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/_custom_op_symbolic_registry.py:46]: Typical usage of documentation from the [Google style guide](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods) should be:

```python
"""A one-line summary of the function

An empty line followed by a more descriptive explanation of the function, example, or code samples

Args:

Returns:
"""
```

I think this is what will work with the Python API documentation tool (Sphinx) as well.

File: orttraining/orttraining/python/training/ortmodule/ortmodule.py [hunk 2]
```diff
@@ -310,4 +310,13 @@ def __getstate__(self):
     def __setstate__(self, state):
         self.__dict__.update(state)
 
-        _utils.reinitialize_ortmodule(self)
+        # Re-register contrib OPs
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/ortmodule.py:313]: I think this change is ok, but what was the reason for moving this code?

**Review by baijumeswani — COMMENTED** [review]
Thanks for making this contribution. I think this was long overdue.

File: orttraining/orttraining/python/training/ortmodule/ortmodule.py (hunk 2 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule/ortmodule.py:313]: Never mind, found the reason in the description. :)

**Review by baijumeswani — DISMISSED** [review]

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 3]
```diff
@@ -200,13 +201,35 @@ def __init__(self, module: _FlattenedModule, debug_options: DebugOptions, fallba
...
+        self._feature_map: List[List[str]] = [
+            ["ATen Executor", "ON", "Dispatch ATen operators to ORT's ATen executor"],
+            [
+                "Cast Propagation",
+                "ON" if self._propagate_cast_ops_level > 0 else "OFF",
+                f"Level {self._propagate_cast_ops_level} enabled",
+            ],
+            ["Custom Function", "ON", "Support custom torch.autograd.Function export and execution"],
+            [
+                "Compute Optimizer",
```
Comment (askhade) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:222]: Does this cover label sparsity optimizations? If not shall we add it here?

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 4]
```diff
@@ -532,14 +564,17 @@ def _enable_conditional_optimizations(
...
         2. Input density observer inspects input tensors and returns sparsity results.
         3. If label or embedding input sparsity is found in sparsity results, graph transformer config is updated to
            enable sparsity-based optimization.
 
         """
-
-        # Enable data sparsity inspection if label sparsity optimization is enabled or user wants to print input density.
+        sparse_opt_status = "ON" if self._enable_sparse_optimizer else "OFF"
+        self._feature_map.append(
+            ["Sparsity Optimizer", sparse_opt_status, "Enable/Disable with env ORTMODULE_ENABLE_SPARSE_OPTIMIZER=1/0"]
```
Comment (askhade) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:575]: can this be moved up near line 213 where other options are added?

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 5]
```diff
@@ -564,13 +599,57 @@ def _enable_conditional_optimizations(
                 # Enable sparsity-based optimization when applicable.
                 if len(label_sparsity_results) > 0:
                     graph_transformer_config.sparse_label_input_names = label_sparsity_results
-                    logger.info(f"Label sparsity based optimization is on for {label_sparsity_results}")
+                    self._logger.info("Label sparsity based optimization is on for %s", label_sparsity_results)
```
Comment (askhade) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:602]: nit: ON
Also, can we added the sparsity percentage? Example: 95%. Not sure if that data is easily accessible here.

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 6]
```diff
@@ -564,13 +599,57 @@ def _enable_conditional_optimizations(
...
+                    self._feature_map.append(
+                        [" - Label Sparsity", sparse_opt_status, "Label inputs: " + str(label_sparsity_results)]
+                    )
 
                 if len(embed_sparsity_results) > 0:
                     graph_transformer_config.sparse_embedding_input_names = embed_sparsity_results
-                    logger.info(f"Embedding sparsity based optimization is on for {embed_sparsity_results}")
+                    self._logger.info("Embedding sparsity based optimization is on for %s", embed_sparsity_results)
```
Comment (askhade) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:609]: same comments as above

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 7]
```diff
@@ -564,13 +599,57 @@ def _enable_conditional_optimizations(
...
+                "Auto Fallback",
+                "ON" if self._fallback_manager.policy is not _FallbackPolicy.FALLBACK_DISABLE else "OFF",
+                "Fallback to PyTorch when encountering unsupported ops",
+            ]
+        )
+
+        mode = "training" if self._export_mode == torch.onnx.TrainingMode.TRAINING else "inference"
+        mode = f"{_logger.LogColor.UNDERLINE}{mode}{_logger.LogColor.ENDC}"
+
+        stat = f"\n\n{_logger.LogColor.HEADER}***** ONNX Runtime (ORTModule) is accelerating your model *****{_logger.LogColor.ENDC}\n\n"
```
Comment (askhade) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:642]: nit: ONNX Runtime Training (ORTModule) accelerates training your model

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py (hunk 3 above)
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:222]: Oh, currently sparsity optimizations is listed as a seperated row. 

I have move sparse opt as a sub-feature of compute optimizer to represent the control scope of those features.

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py (hunk 7 above)
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:642]: ORTModule not only accelerate training, but also evaluating, so I avoided using "training your model" in the title here. I changed to `ONNX Runtime Training (ORTModule) is accelerating your model`

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py (hunk 5 above)
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py]: Have printed the density of input if we plan to enable opt for the input. :)

**Review by baijumeswani — APPROVED** [review]

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 8]
```diff
@@ -532,14 +566,27 @@ def _enable_conditional_optimizations(
...
-        # Enable data sparsity inspection if label sparsity optimization is enabled or user wants to print input density.
+        self._feature_map.extend(
+            [
+                [
+                    "Compute Optimizer",
+                    "ON" if self._enable_compute_optimizer else "OFF",
+                    "Enable/Disable with env ORTMODULE_ENABLE_COMPUTE_OPTIMIZER=1/0",
+                ],
+                [
+                    " -FLOPReduction",
```
[the remark this answers is not in the document]
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:583]: Oh yes, let me remove it in other changes.

**Comment by pengwa** [comment]
Thanks a lot for the review, @askhade  @baijumeswani !
